fc/templates/weblab_model.pyx

```diff
@@ -0,0 +1,347 @@
+
+    # Seems to be unused at the moment
+    #cdef Sundials.N_Vector _parameters
+
+    def __init__(self):
+        self.free_variable_name = "time"
+        self.free_variable = 0.0
+
+        # State values
+        self.state = np.zeros({{ states|length }})
+
+        # Mapping from qualified names to state indices; only for states that
+        # have a variable name.
+        self.state_var_map = {}
+        {%- for state in states %}
+        {%- for name in state.var_names %}
+        self.state_var_map['{{ name }}'] = {{ state.index }}
+        {%- endfor %}
+        {%- endfor %}
+
+        # Initial state
+        self.initial_state = np.zeros({{ states|length }})
+        {%- for state in states %}
+        self.initial_state[{{ state.index }}] = {{ state.initial_value }}
+        {%- endfor %}
+
+        # Mapping of parameter qualified names to parameter array indices
+        self.parameter_map = {}
+        {%- for parameter in parameters %}
+        self.parameter_map['{{ parameter.annotation[1] }}'] = {{ parameter.index }}
+        {%- endfor %}
+
+        # Initial parameter values
+        self.parameters = np.zeros({{ parameters|length }})
+        {%- for parameter in parameters %}
+        self.parameters[{{ parameter.index }}] = {{ parameter.initial_value }}
+        {%- endfor %}
+
+        # Local names of output variables
+        self.output_names = []
+        {%- for output in outputs %}
+        {%- if output.length is none %}
+        self.output_names.append('{{ output.annotation[1] }}')
+        {%- else %}
+        self.output_names.append('{{ output.annotation }}')
+        {%- endif %}
+        {%- endfor %}
+
+        # Create and cache list of arrays, to avoid constant list/array
+        # creation
+        self._outputs = []
+        {%- for output in outputs %}
+        {%- if output.length is none %}
+        self._outputs.append(np.array(0.0))
+        {%- else %}
+        self._outputs.append(np.zeros({{ output.length }}))
+        {%- endif %}
+        {%- endfor %}
+        # TODO Handle vector outputs other than state_variable
+
+        self.state = self.initial_state.copy()
+        self.saved_states = {}
+        self.dirty = False
+        self.indent_level = 0
+        self.associate_with_model(self)
+        #self._parameters = Sundials.N_VMake_Serial(
+        #    len(self.parameters),
+        #    <Sundials.realtype*>(<np.ndarray>self.parameters).data
+        #)
+        self.env = ModelWrapperEnvironment(self)
+
+    #def __dealloc__(self):
+    #    if self._parameters != NULL:
+    #        Sundials.N_VDestroy_Serial(self._parameters)
+
+    def get_environment_map(self):
+        """
+        Get a map from ontology prefix to the environment containing model
+        variables annotated with that ontology.
+
+        See :meth:`fc.simulations.AbstractOdeModel.get_environment_map()`.
+        """
+        # TODO Some part of this might need to be generated
```

Comment:
Yeah, eventually it all should be, based on ontologies actually used in the protocol.